server/tournament_service.py

```diff
@@ -0,0 +1,282 @@
+                    "request_id": message.correlation_id,
+                    "error_code": "OTHER"
+                },
+            )
+
+    async def _process_create_game(self, message):
+        body = message.body
+        body = json.loads(body)
+        game = TournamentGame(body)
+        assert game.request_id == message.correlation_id
+        if not await self._fetch_players(game):
+            await self._remove_and_cancel_game(game, make_idle=False)
+            return
+        _notify_players(game)
+
+    async def _fetch_players(self, game):
+        if not game.participants:
+            self._logger.warning("Tournament game requested with empty player list")
+            return False
+        await self.add_tournament_game(game)
+        for participant in game.participants:
+            player_id = participant["player_id"]
+            player = self.player_service[player_id]
+            if player is None:
+                self._logger.warning("Tournament game requested with player id(%s) that could not be found", player_id)
+                game.state = TournamentGameState.PLAYER_NOT_ONLINE
+                game.players_causing_cancel.add(player_id)
+                continue
+            if not player.state == PlayerState.IDLE:
+                self._logger.warning("Tournament game requested with player id(%s), player not idle", player_id)
+                game.state = TournamentGameState.PLAYER_NOT_IDLE
+                game.players_causing_cancel.add(player_id)
+            game.players.append(player)
+        return game.state == TournamentGameState.SCHEDULED
+
+    async def add_tournament_game(self, game):
+        self._games.add(game)
+
+    async def on_is_ready_response(self, message, player):
+        game: TournamentGame = await self._get_game_for_request_id(message["request_id"])
+        if game is None:
+            await player.lobby_connection.send_warning("You try to ready up for a game that does not exist")
+            return
+        if game.state != TournamentGameState.CONFIRMATION_PENDING:
+            await player.lobby_connection.send_warning("You try to ready up for a game that is not waiting for "
+                                                       "confirmation")
+            return
+        if player not in game:
+            await player.lobby_connection.send_warning("You try to ready up for a game that you are not in")
+            return
+        if player.id in game.players_ready_ids:
+            return
+        await self.add_player_to_ready_list(game, player)
+
+    async def add_player_to_ready_list(self, game, player):
+        game.players_ready_ids.add(player.id)
+        if game.is_ready():
+            await self._launch(game)
+
+    async def _get_game_for_request_id(self, request_id):
+        for game in self._games.copy():
+            if game.request_id == request_id:
+                return game
+
+    async def _launch(self, tournament_game: TournamentGame):
```

Comment:
This looks like a rather large copy/paste. We should try to refactor this.

Reply:
ok